# Capstone — mirrors your deployed research paper

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Question

*The research question and the decision it supports.*

**What performance archetypes exist across a content inventory, and how should a content team
prioritize action on them?**

This supports one decision: which pages a content strategist or editor should review first in a
given sprint — protect, improve, rewrite, monitor — using only structured, safe metrics (traffic,
engagement, freshness, token counts). No article text was used, so this is behavioral clustering,
not semantic clustering. The work is decision-support for a human reviewer, not an automated
content system.

## 2. Data

*Which release, which tables, date windows, what you excluded and why. Public-safe.*

## Data

**Source:** FlyRank ML Internship starter dataset — `data/raw/content_refresh_anonymized.csv`,
30,000 rows, one row per pseudonymized content page, 32 clients, trailing-90-day metrics.

**Date window:** trailing 90 days from extract date, with layered sub-windows (`*_last_30d`,
`*_prev_30d` for trend calculation).

**Excluded fields, with why:**
- `trend_direction` / `trend_pct` as ML **features** (though used as a rule input for the baseline)
  — these are derived the same way a label would be, so using them as model inputs would be circular.
- `provider_used` / `model_used` — production/tooling metadata, not performance signal; mixing it in
  risks clusters reflecting the content pipeline rather than audience behavior.
- `content_id` / `client_id` — identifiers, used only for grouping/splitting, never as features.
- No client names, raw URLs, or query strings appear anywhere in this dataset or this paper —
  all identifiers are pseudonyms per the dataset's own design.

In [1]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/JoeSherif97/FlyRankerML/"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

In [2]:
print(f"Rows: {len(df)} | Unique clients: {df['client_id'].nunique()} | Unique pages: {df['content_id'].nunique()}")

Rows: 30000 | Unique clients: 32 | Unique pages: 30000


## 3. Methodology

*Assumptions, features, label definition, baseline, validation design, leakage checks.*

**Task type:** Clustering (unsupervised) — "what kinds of pages exist," not "will this page
decline." No observed future outcome exists in this snapshot, so classification was not honest here.

**Label:** none. Archetype names (champion, hidden gem, weak/no-demand) are assigned to clusters
*after* inspecting their contents, not defined in advance.

**Baseline:** a transparent, human-readable rule (`stale_declining_visible`,
`declining_visible_fresh`, `stale_visible_stable`, `low_priority`), scored as
`declining × visible × (1 + stale) × impressions_90d`, with one reason code per page.

**Features:** structured 90-day metrics (search economics, performance, engagement, freshness) plus
`has_-flags` for four fields confirmed to be missing by `content_type` rather than randomly
(`search_volume`, `competition`, `cpc`, `word_count`).

**Validation design:** grouped split by `client_id` (`GroupShuffleSplit`), not random — client sizes
are highly skewed (largest client = 23% of rows), so a random split lets the model memorize a
client's own scale rather than generalize.

**Leakage checks:** full checklist run (label-derived fields, product-decision flags, ID columns,
group overlap) plus a deliberate-injection test — adding a known-leaky field
(`trend_pct`) measurably raised holdout silhouette (0.278 → 0.293), confirming the audit harness
was actually sensitive to leakage rather than rubber-stamping a clean-looking list.

## 4. Results (vs baseline)

*Model vs baseline on the same split. The honest table.*

In [3]:
comparison_table = pd.DataFrame({
    "Metric": [
        "Groups found",
        "Resolution inside baseline's largest bucket (21,085 rows)",
        "Holdout silhouette — random split",
        "Holdout silhouette — grouped-by-client split (honest)",
        "Holdout silhouette — with deliberately leaked feature",
    ],
    "Baseline (4 hand-written buckets)": ["4", "1 undifferentiated bucket", "—", "—", "—"],
    "K-Means (k=4)": ["4", "4 distinct sub-clusters", "0.294", "0.278", "0.293 (confirms leakage detection works)"],
})
print(comparison_table.to_string(index=False))

                                                   Metric Baseline (4 hand-written buckets)                            K-Means (k=4)
                                             Groups found                                 4                                        4
Resolution inside baseline's largest bucket (21,085 rows)         1 undifferentiated bucket                  4 distinct sub-clusters
                        Holdout silhouette — random split                                 —                                    0.294
    Holdout silhouette — grouped-by-client split (honest)                                 —                                    0.278
    Holdout silhouette — with deliberately leaked feature                                 — 0.293 (confirms leakage detection works)


K-Means found real but partial structure beyond the baseline's four buckets — most clearly a
382-page high-performing cluster (median 503 sessions, 2.63% engagement, page-1 position) and a
2,469-page near-invisible cluster (median 6 impressions). However, a shallow decision tree reading
the clusters back showed the split leaned heavily on `has_word_count` (importance 0.271) — a
near-proxy for `content_type` — meaning part of the clustering rediscovers known content categories
rather than finding a wholly new behavioral pattern. This is stated as a limitation, not smoothed over.

## 5. Limitations

*What this work cannot claim.*

- **No article text** — clustering is behavioral/structural, not semantic. Archetype names describe
  traffic and engagement shape, not topic or meaning.
- **Content-type confound** — the strongest driver of cluster split (`has_word_count`) is a near-proxy
  for `content_type`, so part of the "archetype" signal is closer to content category than genuine
  behavioral discovery.
- **No validated future outcome** — the baseline rule uses `trend_direction`, itself a defined rule
  (30d vs. prev 30d), not an independently observed future result. No precision@K against a real
  future window exists in this snapshot.
- **Client concentration** — one client holds 23% of rows; the priority queue's top candidates are
  disproportionately drawn from a small number of large clients, confirmed directly rather than assumed.
- **Small holdout for grouped validation** — the grouped split's holdout covers only 7 of 32 clients;
  a single unusual client could swing the reported silhouette more than a larger holdout would.
- **Cross-sectional, one snapshot** — no causal claims are supported. This is observed, directional,
  decision-support structure only.

## 6. Ranked recommendations

*The action playbook output — the paper's recommendations section.*

In [4]:
import os
print("cwd:", os.getcwd())
print("work/ exists:", os.path.isdir("work"))
print("work/outputs/ exists:", os.path.isdir("work/outputs"))
if os.path.isdir("work/outputs"):
    print("contents:", os.listdir("work/outputs"))

cwd: /content/flyrank-ml-internship-starter
work/ exists: True
work/outputs/ exists: True
contents: ['monitoring_baseline.json', 'human_review_sample.csv', 'playbook_summary_by_action.csv', 'content_action_playbook.csv']


In [5]:
# pulled directly from the ML-10 playbook export
playbook = pd.read_csv("work/outputs/content_action_playbook.csv")
print(playbook["action"].value_counts())
print(playbook.groupby("action", observed=True)[["impressions_90d", "priority_score"]].median())

action
improve                14907
monitor                12065
protect_and_refresh     3015
rewrite                   13
Name: count, dtype: int64
                     impressions_90d  priority_score
action                                              
improve                        780.0           656.0
monitor                        447.0             0.0
protect_and_refresh          15342.0           816.0
rewrite                       4556.0          9112.0


1. **Protect and refresh champions first** (3,015 pages) — highest-value pages, often also going
   stale; losing these costs the most.
2. **Rewrite the small, high-confidence stale-and-declining group** (13 pages) — still visible,
   losing ground, untouched 180+ days.
3. **Improve the larger declining-but-fresh group** (14,907 pages) — investigate *why* a recently
   updated page is still declining before assuming a rewrite fixes it.
4. **Monitor the rest** (12,065 pages) — includes weak/no-demand pages where `avg_position` is
   statistically unreliable due to tiny impression bases; do not prune on position alone.

No action here should be automated end-to-end — see the human-review and no-go list in the ML-10
notebook for what a person must check before acting on any recommendation.

## 7. Artifacts the paper embeds

*Generate/collect the charts and tables your deployed page will show.*

In [6]:
import matplotlib.pyplot as plt

# Chart 1: action queue distribution
action_counts = playbook["action"].value_counts()
fig, ax = plt.subplots(figsize=(6,4))
action_counts.plot(kind="bar", ax=ax, color="#2F6F6B")
ax.set_title("Pages per recommended action")
ax.set_ylabel("Page count")
plt.tight_layout()
plt.savefig("work/outputs/fig_action_distribution.png", dpi=150)
plt.close()

# Chart 2: baseline vs clustering resolution inside low_priority
comparison_table.to_csv("work/outputs/table_model_vs_baseline.csv", index=False)

print("Saved: fig_action_distribution.png, table_model_vs_baseline.csv")

Saved: fig_action_distribution.png, table_model_vs_baseline.csv


In [9]:
import os
os.makedirs("docs", exist_ok=True)
print("docs/ exists:", os.path.isdir("docs"))

docs/ exists: True


In [11]:
import shutil
shutil.copy("work/outputs/fig_action_distribution.png", "docs/fig_action_distribution.png")

'docs/fig_action_distribution.png'

In [12]:
%%writefile docs/index.html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Structured Content Archetype Clustering — A Portfolio Lifecycle Study</title>
<link rel="preconnect" href="https://fonts.googleapis.com">
<link href="https://fonts.googleapis.com/css2?family=Fraunces:opsz,wght@9..144,400;9..144,500;9..144,600;9..144,700&family=IBM+Plex+Sans:wght@400;500;600;700&family=IBM+Plex+Mono:wght@400;500;600&display=swap" rel="stylesheet">
<style>
  :root {
    --paper: #F6F4EE;
    --paper-raised: #FBFAF6;
    --ink: #1B231F;
    --ink-soft: #4A544E;
    --growth: #2F6F6B;
    --growth-soft: #E4EEEC;
    --decay: #B5822C;
    --decay-soft: #F5EBD9;
    --rule: #D8D3C5;
    --max-w: 780px;
  }

  * { box-sizing: border-box; }

  html { scroll-behavior: smooth; }

  body {
    margin: 0;
    background: var(--paper);
    color: var(--ink);
    font-family: 'IBM Plex Sans', sans-serif;
    line-height: 1.65;
    font-size: 17px;
  }

  h1, h2, h3 { font-family: 'Fraunces', serif; font-weight: 600; letter-spacing: -0.01em; color: var(--ink); }

  .mono { font-family: 'IBM Plex Mono', monospace; }

  a { color: var(--growth); text-decoration-thickness: 1px; }
  a:hover { color: var(--decay); }

  /* ---------- Lifecycle sparkline signature ---------- */
  .sparkline-divider {
    width: 100%;
    height: 56px;
    margin: 3.5rem 0;
    display: block;
  }

  /* ---------- Layout shell ---------- */
  .wrap {
    max-width: var(--max-w);
    margin: 0 auto;
    padding: 0 1.5rem;
  }

  header.hero {
    padding: 5rem 0 2rem;
    border-bottom: 1px solid var(--rule);
  }

  .eyebrow {
    font-family: 'IBM Plex Mono', monospace;
    font-size: 0.78rem;
    text-transform: uppercase;
    letter-spacing: 0.12em;
    color: var(--growth);
    margin-bottom: 1rem;
  }

  h1.title {
    font-size: clamp(2.1rem, 5vw, 3.1rem);
    line-height: 1.08;
    margin: 0 0 1.25rem;
  }

  .subtitle {
    font-size: 1.15rem;
    color: var(--ink-soft);
    max-width: 640px;
    margin: 0 0 2.5rem;
  }

  .hero-stats {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(140px, 1fr));
    gap: 1.25rem;
    margin-top: 2.5rem;
  }

  .stat {
    border-left: 2px solid var(--growth);
    padding-left: 0.9rem;
  }
  .stat.decay { border-left-color: var(--decay); }

  .stat-num {
    font-family: 'IBM Plex Mono', monospace;
    font-size: 1.5rem;
    font-weight: 600;
    display: block;
  }
  .stat-label {
    font-size: 0.82rem;
    color: var(--ink-soft);
  }

  nav.toc {
    font-family: 'IBM Plex Mono', monospace;
    font-size: 0.85rem;
    padding: 1.5rem 0;
    display: flex;
    flex-wrap: wrap;
    gap: 0.4rem 1.3rem;
    border-bottom: 1px solid var(--rule);
  }
  nav.toc a { color: var(--ink-soft); text-decoration: none; }
  nav.toc a:hover { color: var(--growth); }

  section {
    padding: 3.5rem 0;
    border-bottom: 1px solid var(--rule);
  }
  section:last-of-type { border-bottom: none; }

  .section-num {
    font-family: 'IBM Plex Mono', monospace;
    font-size: 0.8rem;
    color: var(--ink-soft);
    margin-bottom: 0.4rem;
    display: block;
  }

  h2 { font-size: 1.7rem; margin: 0 0 1.3rem; }
  h3 { font-size: 1.15rem; margin: 1.8rem 0 0.8rem; }

  p { margin: 0 0 1.1rem; }

  .callout {
    background: var(--paper-raised);
    border-left: 3px solid var(--growth);
    padding: 1rem 1.3rem;
    border-radius: 2px;
    margin: 1.5rem 0;
    font-size: 0.96rem;
  }
  .callout.warn { border-left-color: var(--decay); }
  .callout strong { display: block; margin-bottom: 0.3rem; font-family: 'IBM Plex Mono', monospace; font-size: 0.78rem; text-transform: uppercase; letter-spacing: 0.08em; }

  table {
    width: 100%;
    border-collapse: collapse;
    font-size: 0.92rem;
    margin: 1.3rem 0;
  }
  th, td {
    text-align: left;
    padding: 0.55rem 0.7rem;
    border-bottom: 1px solid var(--rule);
  }
  th {
    font-family: 'IBM Plex Mono', monospace;
    font-size: 0.75rem;
    text-transform: uppercase;
    letter-spacing: 0.06em;
    color: var(--ink-soft);
    font-weight: 500;
  }
  td.num, th.num { font-family: 'IBM Plex Mono', monospace; text-align: right; }

  .chart-card {
    background: var(--paper-raised);
    border: 1px solid var(--rule);
    border-radius: 4px;
    padding: 1.3rem;
    margin: 1.5rem 0;
  }
  .chart-caption {
    font-size: 0.85rem;
    color: var(--ink-soft);
    margin-top: 0.6rem;
  }

  ul, ol { padding-left: 1.3rem; }
  li { margin-bottom: 0.5rem; }

  .action-list { list-style: none; padding-left: 0; }
  .action-list li {
    display: flex;
    gap: 0.9rem;
    align-items: baseline;
    padding: 0.7rem 0;
    border-bottom: 1px dashed var(--rule);
  }
  .action-rank {
    font-family: 'IBM Plex Mono', monospace;
    color: var(--growth);
    font-weight: 600;
    min-width: 1.6rem;
  }

  footer {
    padding: 3rem 0 5rem;
    font-size: 0.9rem;
    color: var(--ink-soft);
  }
  footer a { color: var(--ink-soft); }

  .badge {
    display: inline-block;
    font-family: 'IBM Plex Mono', monospace;
    font-size: 0.72rem;
    text-transform: uppercase;
    letter-spacing: 0.06em;
    padding: 0.15rem 0.5rem;
    border-radius: 3px;
    background: var(--growth-soft);
    color: var(--growth);
    margin-left: 0.5rem;
  }
  .badge.limit { background: var(--decay-soft); color: var(--decay); }

  @media (max-width: 600px) {
    body { font-size: 16px; }
    header.hero { padding: 3rem 0 1.5rem; }
    section { padding: 2.5rem 0; }
  }
</style>
</head>
<body>

<header class="hero">
  <div class="wrap">
    <div class="eyebrow">Data Report · FlyRank ML Internship Capstone</div>
    <h1 class="title">Structured Content Archetype Clustering</h1>
    <p class="subtitle">
      What performance archetypes exist across a content inventory — and which pages should an
      editor act on first? A behavioral clustering study on 30,000 pages, using only structured
      metrics: no article text, no semantic claims, decision-support only.
    </p>

    <svg class="sparkline-divider" viewBox="0 0 780 56" preserveAspectRatio="none" aria-hidden="true">
      <polyline points="0,40 90,18 180,10 270,14 360,26 450,38 540,46 630,40 690,22 780,12"
        fill="none" stroke="#2F6F6B" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round"/>
      <polyline points="540,46 630,40 690,22 780,12"
        fill="none" stroke="#B5822C" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" stroke-dasharray="1 6"/>
    </svg>

    <div class="hero-stats">
      <div class="stat">
        <span class="stat-num">65.7%</span>
        <span class="stat-label">of sessions from the top 10% of pages</span>
      </div>
      <div class="stat decay">
        <span class="stat-num">54.2%</span>
        <span class="stat-label">of pages trending down over 30 days</span>
      </div>
      <div class="stat">
        <span class="stat-num">26.9%</span>
        <span class="stat-label">low-traffic pages with above-median engagement</span>
      </div>
      <div class="stat decay">
        <span class="stat-num">0.278</span>
        <span class="stat-label">holdout silhouette, honest client-grouped split</span>
      </div>
    </div>
  </div>
</header>

<nav class="toc">
  <div class="wrap" style="display:flex; flex-wrap:wrap; gap:0.4rem 1.3rem;">
    <a href="#abstract">Abstract</a>
    <a href="#intro">Introduction</a>
    <a href="#data">Data</a>
    <a href="#methodology">Methodology</a>
    <a href="#results">Results</a>
    <a href="#limitations">Limitations</a>
    <a href="#recommendations">Recommendations</a>
    <a href="#repro">Reproducibility</a>
    <a href="#credit">Data credit</a>
  </div>
</nav>

<main class="wrap">

  <section id="abstract">
    <span class="section-num">01</span>
    <h2>Abstract</h2>
    <p>
      Content teams managing large page inventories have no systematic way to see which pages
      behave alike, relying instead on gut feel and single-metric sorts. This study used K-Means
      clustering on 30,000 structured content records — 90-day traffic, engagement, and freshness
      metrics, with no article text — to test whether recurring behavioral archetypes exist beyond
      a hand-written priority rule. Clustering found real but partial structure: a small,
      high-value "champion" cluster and a near-invisible "weak-demand" cluster stood out clearly,
      while the remaining two clusters were driven substantially by content type rather than
      independent behavior. A grouped-by-client validation split showed the result was not simply
      client memorization (holdout silhouette 0.278 vs. 0.294 on a random split), and a deliberate
      leakage test confirmed the audit process was genuinely sensitive to contamination. The
      output is a ranked, reason-coded action queue intended for human review — decision-support,
      not an automated recommendation engine, and not a claim about causation or future performance.
    </p>
  </section>

  <section id="intro">
    <span class="section-num">02</span>
    <h2>Introduction</h2>
    <p>
      A content strategist facing a backlog of thousands of pages needs to answer one question
      every sprint: <em>where should limited editor time go first?</em> Single-metric sorts —
      "top 10 by traffic" — miss interaction effects: a low-traffic page can be a hidden gem if
      engagement is high, or genuinely dead weight if it isn't. A wrong call is not free. Pruning a
      quietly-working page wastes rebuild cost later; leaving a declining page alone because it
      still looks strong on one metric compounds lost traffic; rewriting a page that's actually
      competing with a sibling page wastes hours on the wrong fix.
    </p>
    <p>
      This study asks whether unsupervised clustering — grouping pages by structural similarity,
      not by a predicted outcome — can add resolution beyond what a transparent, hand-written rule
      already provides, and states plainly where that added resolution is genuine versus where it
      is closer to rediscovering known content categories.
    </p>
  </section>

  <section id="data">
    <span class="section-num">03</span>
    <h2>Data</h2>
    <p>
      Source: the FlyRank ML Internship starter release — 30,000 pseudonymized content pages
      across 32 clients, trailing-90-day Google Search Console and GA4-derived metrics. No client
      names, raw URLs, or private queries appear anywhere in this dataset; all identifiers are
      pseudonyms by design.
    </p>
    <table>
      <tr><th>Field</th><th>Role</th><th>Why</th></tr>
      <tr><td class="mono">content_id, client_id</td><td>Context</td><td>Grouping and splitting only, never a feature</td></tr>
      <tr><td class="mono">trend_direction, trend_pct</td><td>Excluded as feature</td><td>Derived the same way a label would be — using it risks circularity</td></tr>
      <tr><td class="mono">provider_used, model_used</td><td>Excluded</td><td>Production metadata, not audience-behavior signal</td></tr>
      <tr><td class="mono">15 structured metrics + 4 missingness flags</td><td>Feature</td><td>Safe, structured, knowable at analysis time</td></tr>
    </table>
    <div class="callout">
      <strong>Public-safe note</strong>
      Missingness in <span class="mono">search_volume</span>, <span class="mono">word_count</span>
      and related fields was confirmed to follow <span class="mono">content_type</span> almost
      deterministically (e.g. 100% of one content type structurally lacks keyword-volume data) —
      handled with explicit <span class="mono">has_-flags</span> rather than a blind fill, to avoid
      injecting a fabricated signal.
    </div>
  </section>

  <section id="methodology">
    <span class="section-num">04</span>
    <h2>Methodology</h2>
    <h3>Task framing</h3>
    <p>
      Clustering, not classification: no observed future outcome exists in this snapshot to
      predict, so the honest task is "what kinds of pages exist," evaluated by silhouette score
      plus a human sense-check of whether cluster profiles match recognizable archetypes.
    </p>
    <h3>Baseline</h3>
    <p>
      A transparent, readable rule assigns every page one reason code
      (<span class="mono">stale_declining_visible</span>, <span class="mono">declining_visible_fresh</span>,
      <span class="mono">stale_visible_stable</span>, <span class="mono">low_priority</span>), scored as
      <span class="mono">declining × visible × (1 + stale) × impressions_90d</span> — no fitted weights,
      fully human-auditable.
    </p>
    <h3>Validation design</h3>
    <p>
      Split grouped by <span class="mono">client_id</span>, not random. Client size is highly
      skewed (largest client holds 23% of all rows), so a random split lets the model implicitly
      memorize a client's own scale. A seed-sensitivity check across five seeds confirmed this
      matters: the resulting split ratio swung from 66/34 to 93/7 despite requesting 80/20 every
      time, purely from which large client landed in holdout.
    </p>
    <h3>Leakage audit</h3>
    <p>
      Full checklist run against the final 19-field feature set: no label-derived fields, no
      product-decision flags, no identifier columns, zero client overlap between fit and holdout.
      A deliberate-injection test — adding <span class="mono">trend_pct</span>, a field known to be
      derived the same way as the excluded label — raised holdout silhouette from 0.278 to 0.293,
      confirming the audit harness genuinely detects leakage rather than passing by default.
    </p>
  </section>

  <section id="results">
    <span class="section-num">05</span>
    <h2>Results</h2>
    <div class="chart-card">
      <img src="fig_action_distribution.png" alt="Bar chart: number of pages per recommended action — improve, monitor, protect and refresh, rewrite" style="width:100%; display:block; border-radius:2px;">
      <p class="chart-caption">
        Pages per recommended action, from the final ranked queue. <span class="mono">monitor</span>
        and <span class="mono">improve</span> hold the bulk of the portfolio; the small
        <span class="mono">rewrite</span> bucket (13 pages) is the highest-confidence,
        highest-urgency group — visible, declining, and stale for 180+ days.
      </p>
    </div>

    <table>
      <tr><th>Metric</th><th>Baseline (4 hand-written buckets)</th><th>K-Means (k=4)</th></tr>
      <tr><td>Groups found</td><td class="num">4</td><td class="num">4</td></tr>
      <tr><td>Resolution inside baseline's largest bucket (21,085 rows)</td><td>1 undifferentiated bucket</td><td>Split across all 4 clusters</td></tr>
      <tr><td>Validated metric</td><td>None (rule-based)</td><td class="num">Silhouette = 0.278 (holdout)</td></tr>
    </table>

    <div class="callout warn">
      <strong>Read this carefully</strong>
      A decision tree explaining the clusters found that <span class="mono">has_word_count</span>
      — a near-perfect proxy for <span class="mono">content_type</span> — was the dominant driver
      (importance 0.271) of cluster assignment, ahead of any performance signal. Two of the four
      clusters are genuinely distinct behavioral groups (a 382-page champion cluster and a
      2,469-page weak-demand cluster); the other two are better described as
      "content-type-driven, with performance sub-splits inside," not four independently discovered
      archetypes.
    </div>
  </section>

  <section id="limitations">
    <span class="section-num">06</span>
    <h2>Limitations</h2>
    <ul>
      <li><strong>No article text.</strong> Clustering is behavioral, not semantic — archetype names describe traffic and engagement shape, not topic or meaning.</li>
      <li><strong>Content-type confound.</strong> The strongest cluster-driving feature is a near-proxy for content type, not a purely new behavioral signal.</li>
      <li><strong>No validated future outcome.</strong> The baseline's "declining" label is a defined rule, not an independently observed result — no precision@K against a real future window exists in this snapshot.</li>
      <li><strong>Client concentration.</strong> One client holds 23% of rows; the priority queue's top candidates are disproportionately drawn from a small number of large clients.</li>
      <li><strong>Small grouped holdout.</strong> Only 7 of 32 clients fall in the validation holdout; one unusual client could swing the reported silhouette.</li>
      <li><strong>Cross-sectional data, one snapshot.</strong> No causal claims are supported anywhere in this work — only observed, directional, decision-support structure.</li>
    </ul>
  </section>

  <section id="recommendations">
    <span class="section-num">07</span>
    <h2>Ranked recommendations</h2>
    <p>The output of this work is a reason-coded action queue for human review, not an automated system:</p>
    <ol class="action-list">
      <li><span class="action-rank">01</span><div><strong>Protect and refresh champions first.</strong> The highest-value pages are often also the most neglected relative to their value — losing one costs the most.</div></li>
      <li><span class="action-rank">02</span><div><strong>Rewrite the small, high-confidence stale-and-declining group.</strong> Still visible, losing ground, untouched 180+ days — the clearest actionable signal in the queue.</div></li>
      <li><span class="action-rank">03</span><div><strong>Improve the larger declining-but-fresh group.</strong> A page that's still declining despite a recent update needs investigation, not an assumed rewrite.</div></li>
      <li><span class="action-rank">04</span><div><strong>Monitor the rest.</strong> Includes weak-demand pages where average position is statistically unreliable due to a tiny impression base — do not prune on position alone.</div></li>
    </ol>
    <div class="callout">
      <strong>What should never be automated</strong>
      Auto-publishing or auto-editing content from this queue; auto-pruning or auto-merging pages
      based on a "monitor" label; committing budget from the priority score as if it were a
      validated economic model; or treating a page's cluster/archetype as a permanent identity fed
      into other systems. Every recommendation here requires human review before action.
    </div>
  </section>

  <section id="repro">
    <span class="section-num">08</span>
    <h2>Reproducibility</h2>
    <p>
      All notebooks live in this repository under <span class="mono">work/notebooks/</span>. The
      exported action queue, review sample, and monitoring baseline are in
      <span class="mono">work/outputs/</span>. Random seeds are fixed and stated in each notebook;
      rerunning top-to-bottom reproduces every table and figure on this page.
    </p>
  </section>

  <section id="credit">
    <span class="section-num">09</span>
    <h2>Acknowledgments &amp; data credit</h2>
    <p>
      Built on the FlyRank ML Internship dataset. Data credit and program information:
      <a href="https://flyrank.ai" target="_blank" rel="noopener">flyrank.ai</a>.
    </p>
  </section>

</main>

<footer>
  <div class="wrap">
    <p>Structured Content Archetype Clustering — a decision-support study, not a causal claim. Observed, directional, decision-support only.</p>
  </div>
</footer>

</body>
</html>

Overwriting docs/index.html


In [14]:
!git config --global user.email "youssefsherifwahib@gmail.com"
!git config --global user.name "Youssef Sherif"
!git add docs/
!git commit -m "Add deployed research paper"
!git push

[main bf464cb] Add deployed research paper
 2 files changed, 480 insertions(+)
 create mode 100644 docs/fig_action_distribution.png
 create mode 100644 docs/index.html
fatal: could not read Username for 'https://github.com': No such device or address


In [15]:
from getpass import getpass
token = getpass("Paste your GitHub token: ")

username = "JoeSherif97"
repo_name = "FlyRankerML"

!git remote set-url origin https://{username}:{token}@github.com/{username}/{repo_name}.git
!git push

Paste your GitHub token: ··········
To https://github.com/JoeSherif97/FlyRankerML.git
 ! [rejected]        main -> main (fetch first)
error: failed to push some refs to 'https://github.com/JoeSherif97/FlyRankerML.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.


In [16]:
!git pull --rebase origin main

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 5 (delta 4), reused 0 (delta 0), pack-reused 0 (from 0)
Unpacking objects: 100% (5/5), 1.05 KiB | 540.00 KiB/s, done.
From https://github.com/JoeSherif97/FlyRankerML
 * branch            main       -> FETCH_HEAD
   1bc5e80..175c30c  main       -> origin/main
Successfully rebased and updated refs/heads/main.


In [17]:
!git push

Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 2 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 31.86 KiB | 10.62 MiB/s, done.
Total 5 (delta 2), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
To https://github.com/JoeSherif97/FlyRankerML.git
   175c30c..8e6c2f4  main -> main


In [18]:
url = "https://joesherif97.github.io/FlyRankerML/"  # copy exact URL from Settings → Pages
with open("submission/paper_url.txt", "w") as f:
    f.write(url + "\n")

!git add submission/paper_url.txt
!git commit -m "Add deployed paper URL"
!git push

[main 3c0c6ae] Add deployed paper URL
 1 file changed, 1 insertion(+), 1 deletion(-)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 2 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 406 bytes | 406.00 KiB/s, done.
Total 4 (delta 1), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/JoeSherif97/FlyRankerML.git
   8e6c2f4..3c0c6ae  main -> main


## 5-Minute Demo Outline (Week 8 Showcase)

**Question (30s)**
"FlyRank's own research shows content decays predictably — but which of *your* thousands of
pages needs attention first? I built a page-level triage tool to answer that."

**Method (1 min)**
K-Means clustering (k=4) on 30,000 structured content records — traffic, engagement, freshness,
no article text — validated with a grouped-by-client split (not random), because client sizes
are wildly skewed and a random split would let the model memorize a client's scale instead of
finding a genuinely transferable pattern.

**One chart (1 min)**
Show the holdout silhouette bar chart: 0.294 random split vs. 0.278 grouped split. Talk through
why the honest number is the lower one — the gap is the cost of the random split's optimism, not
noise to explain away.

**One honest result (1.5 min)**
Clustering found two genuinely distinct, actionable groups — a 382-page champion cluster (still
high-performing but going stale) and a 2,469-page weak-demand cluster (near-invisible, where
average position is statistically unreliable and shouldn't drive a prune decision). But a
decision tree explaining the clusters showed the split leaned heavily on `has_word_count` — a
near-proxy for content type — so two of the four clusters are closer to "content-type-driven"
than newly discovered behavior. Say this out loud, don't skip it: it's the finding that shows
the rigor, not a weakness to hide.

**One recommendation (1 min)**
Protect and refresh the champions first — they're the highest-value pages and, per FlyRank's
own decay curve, the ones most expensive to lose if left stale. Show the ranked action queue and
name explicitly what's NOT automated: no auto-publish, no auto-prune, human review required on
every recommendation.

**Close**
"This turns a portfolio-level insight into a page-level decision — decision-support, not a
black box, and I can show you exactly where its resolution runs out."

## Shareable Cuts

### Social post

> Content decays — FlyRank's own research proved that at the portfolio level. I wanted to know
> which pages on a 30,000-page inventory need attention *first*. Built a K-Means clustering
> pipeline validated with a client-grouped split (not random — client sizes are wildly skewed),
> plus a deliberate leakage test to make sure the audit itself actually caught contamination
> when I planted it. Found two genuinely distinct archetypes — a small high-value cluster going
> stale, and a near-invisible weak-demand cluster — and was honest that two of the four clusters
> were closer to "content type" than new behavior. Full paper + reproducible notebooks:
> [your deployed URL]

### Employer-facing summary (3 sentences)

I built a page-level content triage tool that clusters a 30,000-page content inventory into
behavioral archetypes using only structured traffic, engagement, and freshness metrics — no
article text, fully reproducible, validated with a client-grouped split to avoid the model
memorizing individual clients. The clustering surfaced two genuinely distinct, actionable
groups (a small high-value cluster going stale and a near-invisible weak-demand cluster) while
honestly reporting that two of the four clusters were driven more by content type than new
behavior, backed by a deliberate leakage-injection test proving the validation process actually
works. The output is a ranked, human-reviewed action queue — decision-support for a content
team, not an automated system, deployed as a public research page with full methodology and
limitations disclosed.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.